src/lazy/expanded/mod.rs

```diff
@@ -131,84 +134,334 @@ impl<'top, 'data, D: LazyDecoder<'data>> ExpandedStreamItem<'top, 'data, D> {
+                    };
+                    return self.interpret_value(value);
+                }
                 // It's another macro invocation, we'll start evaluating it.
                 EExpression(e_exp) => {
+                    let context = self.context();
+                    let resolved_e_exp = e_exp.resolve(context)?;
+                    // Get the current evaluator or make a new one
+                    let evaluator = match self.evaluator_ptr.get() {
+                        // If there's already an evaluator, dereference the pointer.
+                        Some(ptr) => Self::ptr_to_evaluator(ptr),
+                        // If there's not, make a new one.
+                        None => context
+                            .allocator
+                            // E-expressions always have an empty environment
+                            .alloc_with(move || MacroEvaluator::new(context, Environment::empty())),
+                    };
                     // Push the invocation onto the evaluation stack.
-                    self.evaluator.push(context, e_exp)?;
-                    // Return to the top of the loop to pull the next value (if any) from the evaluator.
-                    continue;
+                    evaluator.push(context, resolved_e_exp)?;
+                    self.evaluator_ptr
+                        .set(Some(Self::evaluator_to_ptr(evaluator)));
+
+                    // Try to get a value by starting to evaluate the e-expression.
+                    if let Some(value) = self.next_from_evaluator()? {
+                        // If we get a value, return it.
+                        return Ok(value);
+                    } else {
+                        // If the expression was equivalent to `(:void)`, return to the top of
+                        // the loop and get the next expression.
+                        continue;
+                    }
                 }
-                EndOfStream => ExpandedStreamItem::EndOfStream,
+                EndOfStream => return Ok(SystemStreamItem::EndOfStream),
             };
-            return Ok(expanded_item);
+        }
+    }
+
+    /// If there is not an evaluation in process, returns `Ok(None)`.
+    /// If there is an evaluation in process but it does not yield another value, returns `Ok(None)`.
+    /// If there is an evaluation in process and it yields another value, returns `Ok(Some(value))`.
+    /// Otherwise, returns `Err`.
+    fn next_from_evaluator<'top>(
+        &'top self,
+    ) -> IonResult<Option<SystemStreamItem<'top, 'data, D>>> {
+        let evaluator_ptr = match self.evaluator_ptr.get() {
+            // There's not currently an evaluator.
+            None => return Ok(None),
+            // There's an evaluator in the process of expanding a macro.
+            Some(ptr) => ptr,
+        };
+        let evaluator = Self::ptr_to_evaluator(evaluator_ptr);
+
+        match evaluator.next(self.context(), 0) {
```

Comment:
What does the `0` represent here? (Maybe add a comment so that it doesn't look like a magic number.)

Reply:
Addressed in [this commit](https://github.com/amazon-ion/ion-rust/pull/674/commits/1db5b308edd4ecdb83f53667f0e4e6fe6dd9ef37).